sky/sdk/lib/widgets/variable_height_scrollable.dart

```diff
@@ -14,18 +12,32 @@ class VariableHeightScrollable extends Scrollable {
   VariableHeightScrollable({
     String key,
     this.builder,
-    this.token
+    this.token,
+    this.layoutState
   }) : super(key: key);
 
   IndexedBuilder builder;
   Object token;
+  BlockViewportLayoutState layoutState;
+
+  // When the token changes the scrollable's contents may have
+  // changed. Remember as much so that after the new contents
+  // have been laid out we can adjust the scrollOffset so that
+  // the last page of content is still visible.
   bool _contentsChanged = true;
 
+  void initState() {
+    assert(layoutState != null);
+    layoutState.addListener(_handleLayoutChanged);
+    super.initState();
+  }
+
   void syncFields(VariableHeightScrollable source) {
     builder = source.builder;
     if (token != source.token)
       _contentsChanged = true;
     token = source.token;
+    assert(layoutState == source.layoutState);
```

Comment:
If you're not allowed to pass a different one each time, this needs to be amply documented and justified.

cc @abarth since I think this may be the first time we've introduced a limitation like this.